### PR TITLE
Fixes #12618 - allow overiding compute profile values

### PR DESCRIPTION
**Pull request by ares**

### CODE REVIEW

File: app/models/host/managed.rb [hunk 1]
```diff
@@ -669,7 +669,7 @@ def set_hostgroup_defaults
 
   def set_compute_attributes
     return unless compute_profile_id && compute_resource_id
-    self.compute_attributes = compute_resource.compute_profile_attributes_for(compute_profile_id)
+    self.compute_attributes = compute_resource.compute_profile_attributes_for(compute_profile_id).merge(self.compute_attributes || {})
```
Comment (domcleal) [inline app/models/host/managed.rb:672]: Is merging the right thing to do, rather than checking if there are already attributes (pre 95d7064)?  It's in progress in PR #2802 so I'm unsure if we should dive in with it here.

That said, I'm not familiar enough with 95d7064 to say whether any of the compute_attributes.empty? can be restored without breaking something else.

**Comment by domcleal** [comment]
I'm still uneasy about the merging, so I looked a bit further into the current behaviour on develop and noticed a couple of things:

1) The compute_profile_id isn't being passed in the parameters hash when the form is serialised for me.  When I select a host group and unpress the "inherit" button for the compute profile, it should be passed, but isn't.

From what I can see, the isExplicit check in 95d70642 is returning false (the data attribute is 'explicit': false when the inherit button is off).

This means the hostgroup compute profile is always in use when the form is submitted or on AJAX requests.

2) set_hostgroup_default checks for `if compute_profile_id_changed?`, which always returns true when creating a new host - since it starts as nil and either gets the compute_profile_id from params, or from the hostgroup.

If it checked whether the compute_profile_id changed _as a result of assign_hostgroup_attributes_, it would be less likely to aggressively overwrite the attributes from the profile.

---

If both of those were fixed, would that solve the problem reliably?  Any time the profile was set by the user (serialised in the form) then we wouldn't overwrite attributes with the hostgroup's profile's attributes.

It would also fix explicitly changing the compute profile on the form, which appears broken to me right now.

**Comment by ShimShtein** [comment]
@domcleal 

Looked into the issue, found the problem. You have mentioned two things earlier:
1. compute_profile_id isn't being passed - It is a good thing, since you want to inherit the profile from the hostgroup.
2. Is right, as you said, for new hosts, compute profile is indeed changed.

What is missing is to check whether the compute attributes were changed explicitly on the form.
Another option would be to split calls from AJAX methods (that will put the attributes explicitly) and calls from submit button (that will just copy values from the params, and will not try to change them at all).

**Comment by domcleal** [comment]
> 1) compute_profile_id isn't being passed - It is a good thing, since you want to inherit the profile from the hostgroup.

Not when "inherit" is unpressed.  The profile should be in the form parameters, but wasn't.

> 2) Is right, as you said, for new hosts, compute profile is indeed changed.

I'm not sure in what sense you mean this "is right".  Sure, the `_changed?` method may be right, but the compute attributes shouldn't be taken from the compute profile if compute_attributes is supplied.

**Comment by ShimShtein** [comment]
> Not when "inherit" is unpressed. The profile should be in the form parameters, but wasn't.

You are right, issue created: http://projects.theforeman.org/issues/12659 it's because we are using `.data` method to set the `data-explicit` value instead of `.attr`

> compute attributes shouldn't be taken from the compute profile if compute_attributes is supplied

The problem is that the compute attributes are always supplied (unless we add some special code there). Probably that means adding the inherit button to the attributes too.

**Comment by domcleal** [comment]
> > compute attributes shouldn't be taken from the compute profile if compute_attributes is supplied
> 
> The problem is that the compute attributes are always supplied (unless we add some special code there).

Yes, that's generally how the New Host form works with compute profiles - the client sends the full set of compute attributes up, whether it's using a compute profile or not.  If the profile changes, the attributes in the form ought to be refreshed so when the form is submitted, the compute attributes from the client reflect what's in the profile (and allow the user to customise it).

The behaviour of overwriting the supplied compute attributes should be limited to when the profile is _changed_.  At the moment the set_hostgroup_defaults validator is overwriting it on every validation (because the check for `compute_profile_id_changed?` is always true.  If this was more specific, so it only overwrote the form's compute attributes when the compute profile ID changed as a result of host group inheritance, it would keep the usual behaviour of trusting the compute attributes parameters.

**Comment by ShimShtein** [comment]
> The behaviour of overwriting the supplied compute attributes should be limited to when the profile is _changed_.

On the server side (and especially in the model layer) I don't know if the value of the compute profile has changed. The value is either present in the `params` (in case the user set it) or it doesn't (in case the user wants to inherit it from the hostgroup). We can remove the automatic call to `#set_compute_attributes`, and call it explicitly in the controller on hostgroup change and compute profile change ajax calls.

**Comment by domcleal** [comment]
> > The behaviour of overwriting the supplied compute attributes should be limited to when the profile is changed.
> 
> On the server side (and especially in the model layer) I don't know if the value of the compute profile has changed. The value is either present in the params (in case the user set it) or it doesn't (in case the user wants to inherit it from the hostgroup).

Perhaps I should change that sentence to:

> The behaviour of overwriting the supplied compute attributes should be limited to when either the compute profile or the host group (when the profile is inherited) are changed?

If inherit is enabled, the parameter's missing when the host group is changed, the attribute gets inherited and so because the attribute is set/changed, we should overwrite compute attributes (as develop does now).

But if inherit is disabled, `params[:compute_profile_id]` is populated when the host group is changed (subject to #2951 of course) and we call Host.new with that profile.  set_hostgroup_defaults is called, it doesn't change compute_profile_id, but in a change from develop, it _shouldn't_ overwrite attributes (currently `compute_profile_id_changed?` would be true so it does).

On a compute profile or resource change, I think the VM partial is reloaded unconditionally, so set_hostgroup_defaults shouldn't be relevant.

In summary:  I think the set_hostgroup_defaults method should only overwrite compute attrs when the compute profile ID was inherited and set in assign_hostgroup_attibutes.

> We can remove the automatic call to #set_compute_attributes, and call it explicitly in the controller on hostgroup change and compute profile change ajax calls.

I think it's OK, the conditional just needs refining.

**Comment by domcleal** [comment]
Disregard my suggestion I think about changing how set_compute_attributes is called from set_hostgroup_defaults, as it would still get called when the form is submitted instead of using the compute attrs from params.  It would need to be in the controller per action, or for changes in the client for which attributes are sent to which actions.

**Comment by ares** [comment]
Closing in favour of #2957
